File: src/main/java/org/aopbuddy/plugin/toolwindow/panel/GroovyConsolePanel.java
package org.aopbuddy.plugin.toolwindow.panel;

import com.intellij.icons.AllIcons;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.ui.OnePixelSplitter;
import com.intellij.ui.border.CustomLineBorder;
import com.intellij.ui.components.JBTabbedPane;
import com.intellij.util.ui.JBUI;
import org.aopbuddy.plugin.infra.ToolWindowUpdateNotifier;
import org.aopbuddy.plugin.service.ConsoleStateService;
import org.aopbuddy.plugin.service.DbSyncService;
import org.aopbuddy.plugin.service.HeartBeatService;
import org.aopbuddy.plugin.service.JvmService;
import org.aopbuddy.plugin.toolwindow.model.*;
import org.aopbuddy.plugin.toolwindow.view.*;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class GroovyConsolePanel extends OnePixelSplitter implements Disposable {

  private static final Logger LOGGER = Logger.getInstance(GroovyConsolePanel.class);

  private final Project project;
  private final ConsoleStateService consoleStateService;
  private final JvmService jvmService;
  private final DbSyncService dbSyncService;
  private final HeartBeatService heartBeatService;

  private final GroovyEditorView groovyEditorView;
  private final ClassloaderView classloaderView;
  private final ClassloaderModel classloaderModel;
  private final RunResultView runResultView;
  private final RunResultModel runResultModel = new RunResultModel();
  private final RecordModel recordModel;
  private final AttachView attachView;
  private final AttachModel attachModel;
  private final RunView runView;
  private final RunModel runModel;


  public GroovyConsolePanel(Project project) {
    super(false, "JZ.ConsoleRun", 0.6F);
    this.project = project;
    this.consoleStateService = project.getService(ConsoleStateService.class);
    this.jvmService = project.getService(JvmService.class);
    this.dbSyncService = project.getService(DbSyncService.class);
    this.heartBeatService = project.getService(HeartBeatService.class);

    this.groovyEditorView = new GroovyEditorView(project);
    this.runResultView = new RunResultView(project, runResultModel);
    this.recordModel = new RecordModel(consoleStateService, dbSyncService);
    this.attachModel = new AttachModel(jvmService, heartBeatService, consoleStateService);
    this.attachView = new AttachView(attachModel);
    this.runModel = new RunModel(jvmService, runResultModel, groovyEditorView);
    this.runView = new RunView(runModel);
    this.classloaderModel = new ClassloaderModel(consoleStateService);
    this.classloaderView = new ClassloaderView(classloaderModel);
    setBorder(BorderFactory.createEmptyBorder());
    setFirstComponent(getGroovyConsolePanel());
    setSecondComponent(getJvmResultInfoPanel());
    project.getMessageBus().connect(this).subscribe(
        ToolWindowUpdateNotifier.GROOVY_CONSOLE_CHANGED_TOPIC,
        (ToolWindowUpdateNotifier) this.groovyEditorView.getGroovyEditor()::setText);
    // 考虑java bean的监听 替代
    project.getMessageBus().connect(this).subscribe(
        ToolWindowUpdateNotifier.ATTACH_STATUS_CHANGED_TOPIC,
        (ToolWindowUpdateNotifier) message -> {
          boolean status = Boolean.parseBoolean(message);
          if (!status) {
            this.dbSyncService.reInit();
            this.heartBeatService.setRunning(false);
            this.classloaderModel.setClassloaders(new ArrayList<>());
            this.classloaderModel.setSelectedItem(null);
            consoleStateService.setIp(null);
          } else {
            this.updateClassloaderComboBox();
          }
          this.attachModel.setStatus(status);
        }
    );
  }

  @Override
  public void dispose() {
    if (attachView != null) {
      attachView.dispose();
    }
  }


  private JComponent getGroovyConsolePanel() {
    JPanel toolbarPanel = new JPanel();
    toolbarPanel.setPreferredSize(new Dimension(-1, 30));
    toolbarPanel.setLayout(new BoxLayout(toolbarPanel, 0));
    toolbarPanel.setBorder(new CustomLineBorder(JBUI.insetsBottom(1)));
    toolbarPanel.add(this.attachView.getAttachButton());
    toolbarPanel.add(this.classloaderView.getClassloaderComboBox());
    toolbarPanel.add(this.runView.getRunButton());
    toolbarPanel.add(createRecordActionButton());
    toolbarPanel.add(Box.createHorizontalGlue());
    toolbarPanel.add(Box.createHorizontalStrut(5));
    JPanel rootPanel = new JPanel(new BorderLayout());
    rootPanel.setBorder(JBUI.Borders.empty());
    rootPanel.add(toolbarPanel, "North");
    rootPanel.add(this.groovyEditorView.getGroovyEditor(), "Center");
    return rootPanel;
  }

  private JComponent getJvmResultInfoPanel() {
    JPanel toolbarPanel = new JPanel();
    // 右侧顶部工具栏
    toolbarPanel.setPreferredSize(new Dimension(-1, 30));
    toolbarPanel.setLayout(new BoxLayout(toolbarPanel, 0));
    toolbarPanel.setBorder(new CustomLineBorder(JBUI.insetsBottom(1)));
    toolbarPanel.add(createClearActionButton());
    toolbarPanel.add(Box.createHorizontalGlue());
//        toolbarPanel.add(Box.createHorizontalStrut(5));
    // 结果面板
    JBTabbedPane groovyTabbedPane = new JBTabbedPane();
    groovyTabbedPane.setBorder(new CustomLineBorder(JBUI.insetsTop(1)));
    groovyTabbedPane.add("执行方法结果", this.runResultView.getRunStatusEditor());
    JPanel rootPanel = new JPanel(new BorderLayout());
    rootPanel.setBorder(JBUI.Borders.empty());
    rootPanel.add(toolbarPanel, "North");
    rootPanel.add(groovyTabbedPane, "Center");
    return rootPanel;
  }


  private void updateClassloaderComboBox() {
    List<String> classloaders = jvmService.getClassloaders();
    classloaders.removeIf(classloader -> classloader.contains("bytebuddy")||classloader.contains("AopAgent"));
    classloaderModel.setClassloaders(classloaders);
    if (!classloaders.isEmpty()) {
      classloaderModel.setSelectedItem(classloaders.get(0));
      consoleStateService.setSelectedClassloader(classloaders.get(0));
    }
  }


  private Component createRecordActionButton() {
    JButton recordActionButton = new JButton();
    recordActionButton.setIcon(AllIcons.CodeWithMe.CwmCamOn);
    recordActionButton.setContentAreaFilled(false);
    recordActionButton.setToolTipText("录制");
    recordActionButton.setPreferredSize(new Dimension(30, -1));
    recordActionButton.addActionListener(e -> {
      RecordFrame instance = RecordFrame.getInstance(this.project);
      instance.showWindow();
    });
    return recordActionButton;
  }

  private Component createClearActionButton() {
    JButton clearActionButton = new JButton();
    clearActionButton.setIcon(AllIcons.Actions.GC);
    clearActionButton.setContentAreaFilled(false);
    clearActionButton.setToolTipText("清除执行结果");
    clearActionButton.setPreferredSize(new Dimension(30, -1));
    clearActionButton.addActionListener(e -> {
      this.runResultModel.setStatus("");
    });
    return clearActionButton;
  }
}

File: src/main/kotlin/org/aopbuddy/plugin/toolwindow/panel/GroovyConsolePanel.kt
package org.aopbuddy.plugin.toolwindow.panel

import com.intellij.icons.AllIcons
import com.intellij.openapi.application.EDT
import com.intellij.openapi.fileTypes.FileTypes
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.ui.OnePixelSplitter
import com.intellij.ui.border.CustomLineBorder
import com.intellij.ui.components.JBTabbedPane
import com.intellij.util.ui.JBUI
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.aopbuddy.plugin.infra.ToolWindowUpdateNotifier
import org.aopbuddy.plugin.infra.config.ServerConfig
import org.aopbuddy.plugin.service.ConsoleStateService
import org.aopbuddy.plugin.service.JvmService
import org.aopbuddy.plugin.toolwindow.component.HintComboBox
import org.aopbuddy.plugin.toolwindow.component.MyEditorTextField
import org.jetbrains.plugins.groovy.GroovyFileType
import java.awt.BorderLayout
import java.awt.Dimension
import javax.swing.*

class GroovyConsolePanel(project: Project) : OnePixelSplitter(false, "JZ.ConsoleRun", 0.6f) {
    private val coroutineScope = CoroutineScope(Dispatchers.EDT)
    private val consoleStateService = project.getService(ConsoleStateService::class.java)
    private val jvmService = JvmService.getInstance(project)
    private val groovyEditor = MyEditorTextField(project, GroovyFileType.GROOVY_FILE_TYPE).apply {
        border = JBUI.Borders.empty(5)
    }
    private val pidComboBox = HintComboBox<String>(200).apply {
        addItem("1.点此添加JVM进程")
        setSelectedIndex(0)
        addActionListener {
            if (selectedIndex == 0) {
                showJvmProcessSelection()
            } else {
                // 更新状态服务中的选中PID
                consoleStateService.serverName = selectedItem.toString()
                updateClassloaderComboBox()
            }
        }
    }
    private val classloaderComboBox = HintComboBox<String>(200).apply {
        setHint("2.选择classLoader")
        addActionListener {
            // 更新状态服务中的选中ClassLoader
            selectedItem?.let {
                consoleStateService.selectedClassloader = it.toString()
            }
        }
    }
    private val runStatusEditor = MyEditorTextField(project, FileTypes.PLAIN_TEXT).apply {
        border = JBUI.Borders.empty(5)
    }
    private val listenerStatusEditor = MyEditorTextField(project, FileTypes.PLAIN_TEXT).apply {
        border = JBUI.Borders.empty(5)
    }

    init {
        firstComponent = getGroovyConsolePanel()
        secondComponent = getJvmResultInfoPanel()
        project.messageBus.connect().subscribe(
            ToolWindowUpdateNotifier.UPDATE_TOPIC,
            ToolWindowUpdateNotifier { text -> groovyEditor.text = text }
        )
    }


    private fun getGroovyConsolePanel(): JComponent {
        val toolbarPanel = JPanel().apply {
            preferredSize = Dimension(-1, 30)
            layout = BoxLayout(this, 0)
            border = CustomLineBorder(JBUI.insetsBottom(1))
            add(pidComboBox)
            add(classloaderComboBox)
            add(createRunButton())
            add(Box.createHorizontalGlue())
            add(Box.createHorizontalStrut(5))
        }

        return JPanel(BorderLayout()).apply {
            border = JBUI.Borders.empty()
            add(toolbarPanel, BorderLayout.NORTH)
            add(groovyEditor, BorderLayout.CENTER)
        }
    }

    private fun getJvmResultInfoPanel(): JComponent {
        val toolbarPanel = JPanel().apply {
            // 右侧顶部工具栏
            preferredSize = Dimension(-1, 30)
            layout = BoxLayout(this, 0)
            border = CustomLineBorder(JBUI.insetsBottom(1))
            add(createClearActionButton())
            add(Box.createHorizontalGlue())
            add(Box.createHorizontalStrut(5))
        }

        val groovyTabbedPane = JBTabbedPane().apply {
            border = CustomLineBorder(JBUI.insetsTop(1))
            addTab("执行方法结果", runStatusEditor)
            addTab("监控方法结果", listenerStatusEditor)
            addChangeListener {
                if (selectedIndex == 1) {
                    coroutineScope.launch {
                        val eval = jvmService.evalAsync("readListenerLog()")
                        listenerStatusEditor.text = eval
                    }
                }
            }
        }

        return JPanel(BorderLayout()).apply {
            border = JBUI.Borders.empty()
            add(toolbarPanel, BorderLayout.NORTH)
            add(groovyTabbedPane, BorderLayout.CENTER)
        }
    }

    private fun showJvmProcessSelection() {
        coroutineScope.launch {
            try {
                val processList = jvmService.getJvmsAsync()
                // 显示选择对话框
                val selectedProcess = JvmProcessSelectorDialog.showAndGet(processList)

                // 如果用户选择了进程
                if (selectedProcess != null) {
                    if (!ServerConfig.getInstance().serverMap.containsKey(selectedProcess)) {
                        try {
                            jvmService.attachAsync(selectedProcess)
                            updatePidComboBox(selectedProcess)
                        } catch (e: Exception) {
                            Messages.showErrorDialog(
                                "连接JVM进程失败: ${e.message}",
                                "连接错误"
                            )
                        }
                    } else {
                        updatePidComboBox(selectedProcess)
                    }
                }
            } catch (e: Exception) {
                Messages.showErrorDialog(
                    "获取JVM进程列表失败: ${e.message}",
                    "获取错误"
                )
            }
        }
    }

    private fun updatePidComboBox(selectedProcess: String) {
        // 检查PID是否已存在于ComboBox中
        var exists = false
        for (i in 0 until pidComboBox.itemCount) {
            val item = pidComboBox.getItemAt(i)
            if (item != null && item == selectedProcess) {
                exists = true
                break
            }
        }
        // 如果不存在，则添加到ComboBox中
        if (!exists) {
            pidComboBox.addItem(selectedProcess)
            pidComboBox.selectedItem = selectedProcess
            // 更新状态服务中的选中PID
            consoleStateService.serverName = selectedProcess
        }
    }

    private fun updateClassloaderComboBox() {
        coroutineScope.launch {
            try {
                val classloaders = jvmService.getClassloadersAsync()
                classloaderComboBox.removeAllItems()
                for (classloader in classloaders) {
                    classloaderComboBox.addItem(classloader)
                }
                if (classloaders.isNotEmpty()) {
                    classloaderComboBox.selectedIndex = 0
                    // 更新状态服务中的可用ClassLoader列表和选中ClassLoader
                    consoleStateService.availableClassloaders = classloaders
                    consoleStateService.selectedClassloader = classloaders[0]
                }
            } catch (e: Exception) {
                Messages.showErrorDialog(
                    "获取类加载器列表失败: ${e.message}",
                    "获取错误"
                )
            }
        }
    }

    private fun createRunButton(): JButton {
        return JButton().apply {
            icon = AllIcons.Actions.RunAll
            isContentAreaFilled = false
            toolTipText = "执行"
            preferredSize = Dimension(30, 30)
            addActionListener {
                coroutineScope.launch {
                    try {
                        val result = jvmService.evalAsync(groovyEditor.text)
                        runStatusEditor.text = result
                    } catch (e: Exception) {
                        Messages.showErrorDialog(
                            "执行脚本失败: ${e.message}",
                            "执行错误"
                        )
                    }
                }
            }
        }
    }

    private fun createClearActionButton(): JButton {
        return JButton().apply {
            icon = AllIcons.Actions.GC
            isContentAreaFilled = false
            toolTipText = "清除执行结果"
            preferredSize = Dimension(30, -1)
            addActionListener {
                runStatusEditor.text = ""
                listenerStatusEditor.text = ""
            }
        }
    }
}